internal/sqldb/csv.go

```diff
@@ -0,0 +1,220 @@
+func listCSVFiles(dir string) ([]*csvHandle, []*csvHandle, error) {
+	var obsCSVs []*csvHandle
+	var tripleCSVs []*csvHandle
+	if bucketName, objectPrefix, ok := parseGCSPath(dir); ok {
+		// Read from GCS
+		ctx := context.Background()
+		client, err := storage.NewClient(ctx)
+		if err != nil {
+			return nil, nil, err
+		}
+		defer client.Close()
+		bucket := client.Bucket(bucketName)
+		query := &storage.Query{
+			Prefix: objectPrefix,
+		}
+		it := bucket.Objects(ctx, query)
+		for {
+			objAttrs, err := it.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				return nil, nil, err
+			}
+			// Check if it's a CSV file
+			if strings.HasSuffix(objAttrs.Name, ".csv") {
+				rc, err := bucket.Object(objAttrs.Name).NewReader(ctx)
+				if err != nil {
+					return nil, nil, err
+				}
+				parts := strings.Split(objAttrs.Name, "/")
+				fileName := parts[len(parts)-1]
+				if strings.HasPrefix(fileName, "observations") {
+					obsCSVs = append(
+						obsCSVs,
+						&csvHandle{
+							f:     rc,
+							name:  objAttrs.Name,
+							close: func() { rc.Close() },
+						},
+					)
+				} else if strings.HasPrefix(fileName, "triples") {
+					tripleCSVs = append(
+						tripleCSVs,
+						&csvHandle{
+							f:     rc,
+							name:  objAttrs.Name,
+							close: func() { rc.Close() },
+						},
+					)
+				}
+			}
+		}
+	} else {
+		// Read from local files
+		files, err := os.ReadDir(dir)
+		if err != nil {
+			return nil, nil, err
+		}
+		for _, file := range files {
+			if fName := file.Name(); strings.HasSuffix(fName, ".csv") {
+				f, err := os.Open(filepath.Join(dir, fName))
+				if err != nil {
+					return nil, nil, err
+				}
+				if strings.HasPrefix(fName, "observations") {
+					obsCSVs = append(
+						obsCSVs,
+						&csvHandle{
+							f:     f,
+							name:  fName,
+							close: func() { f.Close() },
+						},
+					)
+				} else if strings.HasPrefix(fName, "triples") {
+					tripleCSVs = append(
+						tripleCSVs,
+						&csvHandle{
+							f:     f,
+							name:  fName,
+							close: func() { f.Close() },
+						},
+					)
+				}
+			}
+		}
+	}
+	return obsCSVs, tripleCSVs, nil
+}
+
+func processObservationCSV(
+	medatata *resource.Metadata,
+	ch *csvHandle,
+	provID string,
+) (
+	[]*observation,
+	error,
+) {
+	records, err := csv.NewReader(ch.f).ReadAll()
+	if err != nil {
+		return nil, err
+	}
+	numRecords := len(records)
+	if numRecords < 2 {
+		return nil, status.Errorf(codes.FailedPrecondition,
+			"Empty CSV file %s", provID)
+	}
+
+	// Load header.
+	header := records[0]
+	if len(header) < 3 {
+		return nil, status.Errorf(codes.FailedPrecondition,
+			"Less than 3 columns in CSV file %s", provID)
+	}
+	numColumns := len(header)
+
+	// Generate observations.
+	observations := []*observation{}
+	for i := 1; i < numRecords; i++ {
+		record := records[i]
+		for j := 2; j < numColumns; j++ {
+			observations = append(observations, &observation{
```

Comment:
Perhaps skip adding observation if value is empty?

Reply:
Done